### PR TITLE
Next Gen Routing

### PR DESCRIPTION
**Pull request by jdmattern**
This PR includes additions to be able to run t-route with the network and outputs of the Next Generation Water Modeling Framework (ngen). The ngen network is based on the HY_Features Conceptual Model https://docs.opengeospatial.org/is/14-111r6/14-111r6.html. The ngen framework produces output flows at nexuses, which are confluences or points of hydrologic significance. These nexuses are each mapped to a single corresponding downstream segment. These nexus flows are read into t-route at those downstream segments as the qlateral inputs. Not all segments in the network are downstream of a nexus. Those segments not downstream of a nexus are padded with input flow values of zero. There is an addition of src/ngen_routing/src/ngen_main.py where the ngen framework makes an automated call to run t-route.

## Additions
Additions referenced above for additional functionality.

## Removals
Removing previous implementation files of ngen routing.

## Testing

1. Ngen YAML file and input data files added for testing

## Checklist

- [x] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [ ] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [ ] Linux
- [ ] Browser

### Accessibility

- [ ] Keyboard friendly
- [ ] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

### CODE REVIEW

**Comment by hellkite500** [comment]
Since this is targeted to merge to a development branch and now master, I support merging this and working from the development branch to improve the implementation and integration.

**Comment by jdmattern-noaa** [comment]
> Since this is targeted to merge to a development branch and now master, I support merging this and working from the development branch to improve the implementation and integration.

I have a few things going on this week. I plan to make a few updates to this next week, which I think would be best to have before merging to a development branch.

**Comment by hellkite500** [comment]
I'll push some of my changes to your fork for now

File: src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx [hunk 1]
```diff
@@ -1151,6 +1151,7 @@ cpdef object compute_network_structured(
                                    array('l',upstream_ids),
                                    wbody_parameters[wbody_index])
                     )
+                wbody_index += 1
```
Comment (hellkite500) [inline src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx:1154]: @jdmattern-noaa what is the purpose of this increment here?

Comment (jdmattern) [reply src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx:1154]: This is a relic from the previous implementation for indexing reservoirs before the binary_find was added that I accidentally left in. I will remove it. 
https://github.com/NOAA-OWP/t-route/blob/7daa4ffc2db1996e384c3cfbb03eba0424b8e279/src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx#L1091

**Pull request merged**